Description: Print sane error when a fastimport file is incomplete.
Origin: commit, revision id: jelmer@samba.org-20120301211718-3hcascmwxdkt2knz
Author: Jelmer Vernooij <jelmer@samba.org>
Bug: https://launchpad.net/bugs/937972
Last-Update: 2012-03-01
Applied-Upstream: merged in revision 355
X-Bzr-Revision-Id: jelmer@samba.org-20120301211718-3hcascmwxdkt2knz

=== modified file 'NEWS'
Index: unstable/NEWS
===================================================================
--- unstable.orig/NEWS	2013-11-24 17:53:19.352598575 -0500
+++ unstable/NEWS	2013-11-24 17:53:19.344598575 -0500
@@ -4,6 +4,12 @@
 
 .. contents::
 
+Bug fixes
+---------
+
+* Print sane error when a fastimport file is incomplete.
+  (Jelmer Vernooij, #937972)
+
 0.13 2012-02-29
 
 Changes
Index: unstable/cmds.py
===================================================================
--- unstable.orig/cmds.py	2013-11-24 17:53:19.352598575 -0500
+++ unstable/cmds.py	2013-11-24 17:53:19.348598575 -0500
@@ -31,12 +31,17 @@
     :param processor_factory: a callable for creating a processor
     :param user_map: if not None, the file containing the user map.
     """
+    from fastimport.errors import ParsingError
+    from bzrlib.errors import BzrCommandError
     from fastimport import parser
     stream = _get_source_stream(source)
     user_mapper = _get_user_mapper(user_map)
     proc = processor_factory(verbose=verbose, **kwargs)
     p = parser.ImportParser(stream, verbose=verbose, user_mapper=user_mapper)
-    return proc.process(p.iter_commands)
+    try:
+        return proc.process(p.iter_commands)
+    except ParsingError, e:
+        raise BzrCommandError("%d: Parse error: %s" % (e.lineno, e))
 
 
 def _get_source_stream(source):
@@ -309,13 +314,18 @@
     def _generate_info(self, source):
         from cStringIO import StringIO
         from fastimport import parser
+        from fastimport.errors import ParsingError
+        from bzrlib.errors import BzrCommandError
         from bzrlib.plugins.fastimport.processors import info_processor
         stream = _get_source_stream(source)
         output = StringIO()
         try:
             proc = info_processor.InfoProcessor(verbose=True, outf=output)
             p = parser.ImportParser(stream)
-            return_code = proc.process(p.iter_commands)
+            try:
+                return_code = proc.process(p.iter_commands)
+            except ParsingError, e:
+                raise BzrCommandError("%d: Parse error: %s" % (e.lineno, e))
             lines = output.getvalue().splitlines()
         finally:
             output.close()
@@ -442,12 +452,16 @@
                     " a newer python-fastimport to use "
                     "--dont-squash-empty-commits")
 
+        from fastimport.errors import ParsingError
         from fastimport import parser
         stream = _get_source_stream(source)
         user_mapper = _get_user_mapper(user_map)
         proc = filter_processor.FilterProcessor(params=params, verbose=verbose)
         p = parser.ImportParser(stream, verbose=verbose, user_mapper=user_mapper)
-        return proc.process(p.iter_commands)
+        try:
+            return proc.process(p.iter_commands)
+        except ParsingError, e:
+            raise BzrCommandError("%d: Parse error: %s" % (e.lineno, e))
 
 
 class cmd_fast_import_info(Command):
Index: unstable/tests/test_commands.py
===================================================================
--- unstable.orig/tests/test_commands.py	2013-11-24 17:53:19.352598575 -0500
+++ unstable/tests/test_commands.py	2013-11-24 17:53:19.348598575 -0500
@@ -228,6 +228,16 @@
         data = self.run_bzr("fast-import file.fi br")[0]
         self.assertEquals(1, tree.branch.revno())
 
+    def test_missing_bytes(self):
+        self.build_tree_contents([('empty.fi', """
+commit refs/heads/master
+mark :1
+committer
+data 15
+""")])
+        self.make_branch_and_tree("br")
+        self.run_bzr_error(['bzr: ERROR: 4: Parse error: line 4: Command commit is missing section committer\n'], "fast-import empty.fi br")
+
 
 class TestFastImportFilter(ExternalBase):
 
